src/main/java/org/openstreetmap/atlas/geography/atlas/change/Change.java

```diff
@@ -0,0 +1,165 @@
+        this.locationToIndex = new MultiMap<>();
+        this.identifier = CHANGE_IDENTIFIER_FACTORY.getAndIncrement();
+    }
+
+    @Override
+    public Rectangle bounds()
+    {
+        return this.bounds;
+    }
+
+    public Collection<FeatureChange> getFeatureChanges()
+    {
+        return this.indexToFeatureChange.values();
+    }
+
+    public int getIdentifier()
+    {
+        return this.identifier;
+    }
+
+    public String getName()
+    {
+        if (this.name == null)
+        {
+            return String.valueOf(this.getIdentifier());
+        }
+        else
+        {
+            return this.name;
+        }
+    }
+
+    @Override
+    public String toString()
+    {
+        final StringList split = new StringList();
+        final StringBuilder builder = new StringBuilder();
+        this.indexToFeatureChange
+                .forEach((index, featureChange) -> split.add(index + " - " + featureChange));
+        builder.append("[Change:");
+        builder.append(System.lineSeparator());
+        builder.append(split.join(System.lineSeparator()));
+        builder.append(System.lineSeparator());
+        builder.append("]");
+        return builder.toString();
+    }
+
+    public Change withName(final String name)
+    {
+        this.name = name;
+        return this;
+    }
+
+    protected void add(final FeatureChange featureChange)
+    {
+        final long currentIndex = this.indexToFeatureChange.isEmpty() ? 0
+                : this.indexToFeatureChange.lastKey() + 1;
+        final Tuple<ItemType, Long> key = new Tuple<>(featureChange.getItemType(),
+                featureChange.getIdentifier());
+        if (!this.identifierToIndex.containsKey(key))
+        {
+            this.identifierToIndex.put(key, currentIndex);
+        }
+        else
+        {
+            throw new CoreException(
+                    "Change already has a feature change {}. Adding {} is not allowed.",
+                    this.identifierToIndex.get(key), featureChange);
+        }
+        this.indexToFeatureChange.put(currentIndex, featureChange);
+        final AtlasEntity reference = featureChange.getReference();
+        if (reference instanceof LocationItem)
+        {
+            this.locationToIndex.add(((LocationItem) reference).getLocation(), currentIndex);
+        }
+        final Rectangle featureBounds = featureChange.bounds();
+        if (this.bounds != null)
+        {
+            this.bounds = this.bounds.combine(featureBounds);
+        }
+        else
+        {
+            this.bounds = featureBounds;
+        }
+    }
+
+    protected Optional<FeatureChange> changeFor(final ItemType itemType, final Long identifier)
+    {
+        final Tuple<ItemType, Long> key = new Tuple<>(itemType, identifier);
+        if (!this.identifierToIndex.containsKey(key))
+        {
+            return Optional.empty();
+        }
+        return Optional.of(this.indexToFeatureChange.get(this.identifierToIndex.get(key)));
```

Comment:
Thoughts on `ofNullable` instead of `of`? Technically the current implementation of `Change.add` prevents `of` from ever throwing a `NullPointerException`, but I think it's better for `changeFor` to not depend on this behavior. Unless you are thinking it should `NullPointerException` deliberately to signal a malformed `Change` object (ie. one where the `identifierToIndex` and the `indexToFeatureChange` maps are incongruent)?

Reply:
Done!